CharReader: Add StructuredError

martinduffy1 opened the pull request:
The move from Json::Reader to Json::CharReader omitted the ability to get structured error messages from the parser. Clients which want to extract the line and column information have to parse the returned string. Add a new getStructuredErrors() method which exposes the error as structured data instead of a string.

Based on the work from @kylefromkitware and comments from @BillyDonahue on #1281.

ahedberg commented:
Out of curiosity, what's the status of this PR? We'd find this addition useful as well.

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1976,6 +1974,15 @@ void CharReaderBuilder::setDefaults(Json::Value* settings) {
   //! [CharReaderBuilderDefaults]
 }
 
+auto CharReader::getStructuredErrors() const -> std::vector<CharReader::StructuredError> {
```
```suggestion
std::vector<CharReader::StructuredError> CharReader::getStructuredErrors() const {
```

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1976,6 +1974,15 @@ void CharReaderBuilder::setDefaults(Json::Value* settings) {
+    return _impl->getStructuredErrors();
+}
+
+bool CharReader::parse(char const* beginDoc, char const* endDoc, Value* root,
+                     String* errs) {
```
```suggestion
bool CharReader::parse(char const* beginDoc, char const* endDoc, Value* root,
                       String* errs) {
```

BillyDonahue commented:
> Out of curiosity, what's the status of this PR? We'd find this addition useful as well.

I have some pending comments that I never hit "Send" on. Sorry about that.

martinduffy1 commented on include/json/reader.h:
```diff
@@ -337,7 +363,7 @@ class JSON_API CharReaderBuilder : public CharReader::Factory {
   ~CharReaderBuilder() override;
 
-  CharReader* newCharReader() const override;
+  CharReader::Impl* newCharReader() const override;
```
I think I'm not understanding your suggestion about this here (https://github.com/open-source-parsers/jsoncpp/pull/1281#issuecomment-816927120)

> Close but I would not do `dynamic_cast` of `this`, but rather hold a polymorphic `_impl`. `OurCharReader` would implement a polymorphic interface and implement `getStructuredErrors()` directly.
> 
> So it would be:
> 
> ```c++
> CharReader {
> public:
>     std::vector<StructuredError> getStructuredErrors() const;
> 
>     /** apologetically virtual only for ABI compatibility. Not an extension point, so `final`. */
>     virtual bool parse(etc...) final;
> 
> private:
>     class Impl;  // Privately provides the polymorphic interface that CharReader provides today.
>     std::unique_ptr<Impl> _impl;
> };
> 
> // In cpp file to prevent inlining and control visibility.
> 
> class CharReader::Impl {
> public:
>     virtual ~Impl() = default;
>     virtual bool parse(etc...) = 0;
>     virtual std::vector<StructuredError> getStructuredErrors() const = 0;
> };
> 
> auto CharReader::getStructuredErrors() const -> std::vector<StructuredError> {
>     return _impl->getStructuredErrors();
> }
> 
> bool CharReader::parse(etc...) {
>     return _impl->parse(etc...);
> }
> ```

This particular change was part of a series of changes related to making `OurCharReader` override `CharReader::Impl`. I did this since you suggested `OurCharReader` implement `getStructuredErrors()` while `CharReader::getStructuredErrors()` calls `_impl->getStructuredErrors()`. Can you explain how you envisioned `_impl` getting the implementation of `getStructuredErrors()` in your suggestion for this?

martinduffy1 replied:
@BillyDonahue can you get back to me about this question so I can finish resolving your comments?

BillyDonahue replied:
Yes sorry for the delay.

I wasn't sure what the question meant by "override" of `CharReader::Impl`, so I was waiting until I had a good amount of time to spend on it. Okay.

I'm thinking that `newCharReader` just returns `CharReader*` as it always has.

ABI-wise, no client code depends on the internals of `CharReader`.  They only ever work with it by pointer and by calling functions on those pointers that live in the JsonCpp library.

But the `newCharReader` function can place whatever kind of `Impl` it wants into the `CharReader` it returns. This hides the details from the callers.

It would make an `Impl` subclass that has a `getStructuredErrors` appropriate for a `CharReader`, and place that into the `CharReader` to which it returns a pointer.

martinduffy1 commented:
@BillyDonahue I responded to all the comments from your last review. Let me know if you have any questions.

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1873,20 +1868,36 @@ std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
-    if (errs) {
-      *errs = reader_.getFormattedErrorMessages();
+      : collectComments_(collectComments), reader_(features) {
+        this->_impl = std::unique_ptr<OurCharReader::Impl>(new OurCharReader::Impl(collectComments, features));
```
If you're in the OurCharReader class scope you don't need to say OurCharReader here.
You can initialize _impl just like you initialized the other members, in the initialization list.

martinduffy1 replied:
You cannot initialize base class members from the initialization list.

BillyDonahue commented on include/json/reader.h:
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
+  std::vector<StructuredError> getStructuredErrors() const;
+
+  class Impl {
+    public:
```
Use surrounding indentation style.

BillyDonahue commented on include/json/reader.h:
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
-                     String* errs) = 0;
+                     String* errs) final;
+
+   /** \brief Returns a vector of structured errors encountered while parsing.
```
This doc should answer some questions about the way CharReader stores the errors after a parse.

Does each parse call reset the stored errors vector?
Does a call to this function consume and reset the errors?

BillyDonahue commented on include/json/reader.h:
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
    */
   virtual bool parse(char const* beginDoc, char const* endDoc, Value* root,
-                     String* errs) = 0;
+                     String* errs) final;
```
remove `final`

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1873,20 +1868,36 @@ std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
-    return ok;
-  }
+
+    std::vector<CharReader::StructuredError> getStructuredErrors() const {
```
add `override`

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1873,20 +1868,36 @@ std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
+
+  class Impl : public CharReader::Impl {
+  public:
+    bool const collectComments_;
```
I think the convention here is for data members to be placed at the bottom of the class def

BillyDonahue replied:
Ok, this one wasn't done. Can you please move these data members to the bottom of the class, after the member functions?

martinduffy1 commented on include/json/reader.h:
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
   }; // Factory
+
+protected:
+  std::unique_ptr<Impl> _impl;
```
It needs to be protected because it is set from OurCharReader

BillyDonahue replied:
`CharReader` could have a protected constructor that accepts a `std::unique_ptr<Impl>` and moves it into the `_impl` data member. `OurCharReader` could call that constructor.

martinduffy1 commented:
Thank you for the quick review @BillyDonahue 
I added some comments about why I am handling impl_ the way I am, and fixed everything else according to your suggestions.

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
-      *errs = reader_.getFormattedErrorMessages();
+      : collectComments_(collectComments), reader_(features) {
+        this->_impl=std::unique_ptr<Impl>(new Impl(collectComments, features));
+      }
```
fix indent

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
-    if (errs) {
-      *errs = reader_.getFormattedErrorMessages();
+      : collectComments_(collectComments), reader_(features) {
+        this->_impl=std::unique_ptr<Impl>(new Impl(collectComments, features));
```
Spaces around the `=` here.
But I think this assignment can be eliminated by calling a protected CharReader constructor.

BillyDonahue replied:
(line 1871) If the `impl` is recording the `collectComments`, then `OurCharReader` shouldn't have a redundant copy of it.  I think `collectComments_` and `reader_` can be removed from the `OurCharReader` class.

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
+
+  class Impl : public CharReader::Impl {
+    bool const collectComments_;
+    OurReader reader_;
```
On Line 1872, `OurCharReader` has an `OurReader` data member.
Isn't this `reader_` redundant with that one?

BillyDonahue replied:
not addressed, for example.

martinduffy1 replied:
I addressed this by removing the OurReader data member from OurCharReader

BillyDonahue commented on include/json/reader.h:
```diff
@@ -262,7 +268,22 @@ class JSON_API CharReader {
+  */
+  std::vector<StructuredError> getStructuredErrors() const;
+
+  class Impl {
```
I think this should be `protected` API, not `public`.

Derived classes need to know what an `Impl` is but nobody else does.

martinduffy1 commented:
@BillyDonahue thanks for your review, I've addressed these comments.

BillyDonahue commented:
> @BillyDonahue thanks for your review, I've addressed these comments.

I still see some comments unaddressed.

martinduffy1 commented:
Which comments are unaddressed? Is the problem that I didn't click resolve
on the comments?
I did notice another indentation error though which I will commit now.

On Tue, Oct 11, 2022 at 5:23 AM Billy Donahue ***@***.***>
wrote:

> @BillyDonahue <https://github.com/BillyDonahue> thanks for your review,
> I've addressed these comments.
>
> I still see some comments unaddressed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/open-source-parsers/jsoncpp/pull/1409#issuecomment-1274390633>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZKCO2OVQIT6JHRG7X37HNDWCUWYVANCNFSM5YAQMQ4Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

BillyDonahue commented:
> Which comments are unaddressed? Is the problem that I didn't click resolve on the comments? I did notice another indentation error though which I will commit now.
> […](#)
> On Tue, Oct 11, 2022 at 5:23 AM Billy Donahue ***@***.***> wrote: @BillyDonahue <https://github.com/BillyDonahue> thanks for your review, I've addressed these comments. I still see some comments unaddressed. — Reply to this email directly, view it on GitHub <[#1409 (comment)](https://github.com/open-source-parsers/jsoncpp/pull/1409#issuecomment-1274390633)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AZKCO2OVQIT6JHRG7X37HNDWCUWYVANCNFSM5YAQMQ4Q> . You are receiving this because you authored the thread.Message ID: ***@***.***>

martinduffy1 commented on src/lib_json/json_reader.cpp:
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
+  public:
+    Impl(bool collectComments, OurFeatures const& features)
+        : collectComments_(collectComments), reader_(features) {}
+    bool parse(char const* beginDoc, char const* endDoc, Value* root,
```
This isn't showing as Outdated, but I added blank lines between these.

martinduffy1 commented:
> > Which comments are unaddressed? Is the problem that I didn't click resolve on the comments? I did notice another indentation error though which I will commit now.
> 
> Please look through some of those comments. I think you'll see what I mean.

The comment you responded to was addressed elsewhere in the code. I added some comments to the issues not showing as Outdated to explain how these were addressed. In the future, I'll make sure to include a description of how the comments were addressed to make it more straightforward to review.

martinduffy1 commented:
> > > Which comments are unaddressed? Is the problem that I didn't click resolve on the comments? I did notice another indentation error though which I will commit now.
> > 
> > 
> > Please look through some of those comments. I think you'll see what I mean.
> 
> The comment you responded to was addressed elsewhere in the code. I added some comments to the issues not showing as Outdated to explain how these were addressed. In the future, I'll make sure to include a description of how the comments were addressed to make it more straightforward to review.

@BillyDonahue can you take a look at this comment and let me know if there are still changes you are waiting on from me, or if you just haven't gotten a chance to review them yet?

BillyDonahue commented:
> > > > Which comments are unaddressed? Is the problem that I didn't click resolve on the comments? I did notice another indentation error though which I will commit now.
> > > 
> > > 
> > > Please look through some of those comments. I think you'll see what I mean.
> > 
> > 
> > The comment you responded to was addressed elsewhere in the code. I added some comments to the issues not showing as Outdated to explain how these were addressed. In the future, I'll make sure to include a description of how the comments were addressed to make it more straightforward to review.
> 
> @BillyDonahue can you take a look at this comment and let me know if there are still changes you are waiting on from me, or if you just haven't gotten a chance to review them yet?

Sorry it's very hard to follow Github PR reviews, so if something is addressed elsewhere I may not make the connection. I will take another deeper look.

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1860,10 +1855,10 @@ String OurReader::getFormattedErrorMessages() const {
 
-std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
-  std::vector<OurReader::StructuredError> allErrors;
+std::vector<CharReader::StructuredError> OurReader::getStructuredErrors() const {
```
The api before this change had an OurReader::getStructuredErrors that returned a std::vector<OurReader::StructuredError>. But now the return type has changed its name, which is an API change, which was I thought what we were trying to avoid in the original proposal by not changing the offsets of the virtual functions in the CharReader class.

martinduffy1 replied:
I've moved those data members and made them private.

I resolved the api change by adding a transformation in OurImpl::getStructuredErrors to convert the CharReader vector to an OurReader vector.

martinduffy1 replied:
@BillyDonahue can you let me know when you get a chance to take a look at this?

martinduffy1 commented:
@BillyDonahue let me know if these changes address your comment

martinduffy1 commented:
@BillyDonahue it's been 8 weeks since your last update on this. Is there anything you're waiting on from me?

BillyDonahue commented:
> @BillyDonahue it's been 8 weeks since your last update on this. Is there anything you're waiting on from me?

Sorry. I made a few suggestions and didn't hear back. If you have resolved them can you comment and resolve?

martinduffy1 commented:
@BillyDonahue I see this marked as approved. Is there a release timeline that I can expect for these changes to get merged?

BillyDonahue commented:
> @BillyDonahue I see this marked as approved. Is there a release timeline that I can expect for these changes to get merged?

Sorry I thought the approval was sufficient for you to merge it rather than me.
I see that we have to rerun Travis.

BillyDonahue commented:
Hm. I don't actually know how to restart the stuck Travis CI job.
I poked around and didn't find anything.
If you push an update maybe it will kickstart itself.

martinduffy1 commented:
@BillyDonahue I squashed the commits and pushed to try to retrigger the CI job. I don't think I can merge this as I don't have write access to the repository. I'm also wondering if there can be a minor version increase when this and #1412 are merged so that we can detect versions with these updates.
